pkg/runtime/runtime.go

```diff
@@ -302,21 +288,55 @@ func (a *DaprRuntime) initRuntime(opts *runtimeOpts) error {
 	a.populateSecretsConfiguration()
 	// Create and start internal and external gRPC servers
 	grpcAPI := a.getGRPCAPI()
+
 	err = a.startGRPCAPIServer(grpcAPI, a.runtimeConfig.APIGRPCPort)
 	if err != nil {
 		log.Fatalf("failed to start API gRPC server: %s", err)
 	}
 	log.Infof("API gRPC server is running on port %v", a.runtimeConfig.APIGRPCPort)
+	a.grpcExternalAPI = grpcAPI
 
+	// Start HTTP Server
+	a.startHTTPServer(a.runtimeConfig.HTTPPort, a.runtimeConfig.ProfilePort, a.runtimeConfig.AllowedOrigins, pipeline)
+	log.Infof("http server is running on port %v", a.runtimeConfig.HTTPPort)
+
+	grpcAPI = a.getGRPCAPI()
 	err = a.startGRPCInternalServer(grpcAPI, a.runtimeConfig.InternalGRPCPort)
 	if err != nil {
 		log.Fatalf("failed to start internal gRPC server: %s", err)
 	}
 	log.Infof("internal gRPC server is running on port %v", a.runtimeConfig.InternalGRPCPort)
+	a.grpcInternalAPI = grpcAPI
 
-	// Start HTTP Server
-	a.startHTTPServer(a.runtimeConfig.HTTPPort, a.runtimeConfig.ProfilePort, a.runtimeConfig.AllowedOrigins, pipeline)
-	log.Infof("http server is running on port %v", a.runtimeConfig.HTTPPort)
+	a.blockUntilAppIsReady()
+
+	err = a.createAppChannel()
+	if err != nil {
+		log.Warnf("failed to open %s channel to app: %s", string(a.runtimeConfig.ApplicationProtocol), err)
+	}
+	a.daprHTTPAPI.SetAppChannel(a.appChannel)
+	a.grpcExternalAPI.SetAppChannel(a.appChannel)
+	a.grpcInternalAPI.SetAppChannel(a.appChannel)
+
+	a.loadAppConfiguration()
+
+	a.initDirectMessaging(a.nameResolver)
+
+	a.daprHTTPAPI.SetDirectMessaging(a.directMessaging)
+	a.grpcExternalAPI.SetDirectMessaging(a.directMessaging)
```

Comment:
please use `grpcAPI.SetDirectMessaging(xxx)` instead of using grpcExternalAPI and InternalAPI. spliting grpc server is the separate story.